### ISSUE TITLE
libsel4: Generated invocation functions don't reflect the entire return type for errors

### ISSUE THREAD

**Issue by kent-mcleod** [seL4/seL4#1321]
Most kernel invocations can return errors. For several types of errors, the kernel actually returns an error message with more information about the error. EG for a seL4_RangeError, the kernel provides rangeErrorMin and rangeErrorMax in the return message.
However libsel4 only declares that invocation functions return seL4_Error which is an enum. The caller is still able to extract the error message out of the message registers themselves, and there are definitions for the payload format of these error provided in libsel4.
Considering that the generated invocation functions take responsibility for marshalling and unmarshalling the call into message registers for the general case, it's likely an oversight that they don't do the same for error cases.

**Comment by kent-mcleod** [comment]
I have a hypothesis that this omission leads people to assume that the kernel only provides error codes (like POSIX syscalls) and we under utilize the ability to encode error information in return messages.

**Comment by lsf37** [comment]
True, if you're only looking at libsel4 types, you'd most likely miss that.

Are you suggesting we should be returning a struct or tagged union instead or that we should add something like a `getErrorInfo` function that returns a struct?

**Comment by Indanz** [comment]
> Considering that the generated invocation functions take responsibility for marshalling and unmarshalling the call into message registers for the general case, it's likely an oversight that they don't do the same for error cases.

The difference is that the general case happens at every call, so marshalling code can be generated. The error case only happens sometime, so it can't be reflected in the return type and be marshalled that way.

The "Error Codes" subchapter of "seL4 API Reference" does define the message format for e.g. seL4_RangeError. Not sure how you can make it clearer to the user in libsel4 itself though.

**Comment by kent-mcleod** [comment]
> The difference is that the general case happens at every call, so marshalling code can be generated. The error case only happens sometime, so it can't be reflected in the return type and be marshalled that way.

The code for unmarshalling the error registers is already generated, but it only places them into the IPC buffer so that a followup call to `seL4_getFault(seL4_MessageInfo_t tag)` will return the fault by value: `seL4_Fault_t`. The fault type is very big on MCS because it has to be large enough to hold a full TCB register context. 

The generated invocation functions could set an output pointer to the IPC buffer address that holds the fault registers, then the caller could choose whether to make a copy or operate on the fault in place.

**Comment by Indanz** [comment]
Oh wow, I didn't know about `seL4_getFault`. That doesn't seem like a very convenient way of decoding errors.
